File: implementations/go/connections/derive.go
package connections

import (
	"strconv"

	"github.com/bpprotocol/blockparty/implementations/go/crypto"
	"github.com/bpprotocol/blockparty/implementations/go/derive"
)

// connectInfo / audienceInfo are the per-epoch HKDF info strings.
func connectInfo(epoch uint64) string {
	return "bpprotocol.org/v1/connect:epoch=" + strconv.FormatUint(epoch, 10)
}

func audienceInfo(epoch uint64) string {
	return "bpprotocol.org/v1/connect-audience:epoch=" + strconv.FormatUint(epoch, 10)
}

// connectionSecret0 derives the epoch-0 connection secret from the two KEM
// shared secrets and the two handshake nonces. Both parties hold ss1, ss2,
// nonce, and nonceResp, so both derive the same secret.
func connectionSecret0(ss1, ss2, nonce, nonceResp []byte) []byte {
	ikm := concat(ss1, ss2)
	salt := concat(nonce, nonceResp)
	return crypto.HKDFSHA256(ikm, salt, []byte(connectInfo(0)), 32)
}

// rotatedSecret derives the next epoch's connection secret by ratcheting the
// previous secret with fresh KEM entropy ss3, salted by the rotation
// ciphertext ct3 (see connections.md §Rotation).
func rotatedSecret(prev, ss3, ct3 []byte, epoch uint64) []byte {
	return crypto.HKDFSHA256(concat(prev, ss3), ct3, []byte(connectInfo(epoch)), 32)
}

// audienceSecret derives the private audience secret for an epoch from the
// connection secret. This is the secret fed to the AEAD layer.
func audienceSecret(connSecret []byte, epoch uint64) []byte {
	return crypto.HKDFSHA256(connSecret, []byte{}, []byte(audienceInfo(epoch)), 32)
}

// audienceCode derives the private audience code for an epoch. It is derived
// from the secret, so it is unlinkable to either participant's address.
func audienceCode(connSecret []byte, epoch uint64) derive.Code {
	return derive.Code(crypto.Keccak256(audienceSecret(connSecret, epoch))[:derive.CodeBytes])
}

func concat(a, b []byte) []byte {
	out := make([]byte, 0, len(a)+len(b))
	out = append(out, a...)
	return append(out, b...)
}

File: js/packages/sdk/src/connections/derive.ts
import { hkdfSha256, keccak256 } from "../crypto/index.js";
import { CODE_BYTES } from "../derive/index.js";
import { concatBytes, utf8ToBytes } from "../bytes.js";

const connectInfo = (epoch: number): string => `bpprotocol.org/v1/connect:epoch=${epoch}`;
const audienceInfo = (epoch: number): string => `bpprotocol.org/v1/connect-audience:epoch=${epoch}`;

const EMPTY_SALT = new Uint8Array(0);

/**
 * The epoch-0 connection secret from the two KEM shared secrets and the two
 * handshake nonces. Both parties hold ss1, ss2, nonce, and nonceResp, so both
 * derive the same secret.
 */
export function connectionSecret0(
  ss1: Uint8Array,
  ss2: Uint8Array,
  nonce: Uint8Array,
  nonceResp: Uint8Array,
): Uint8Array {
  return hkdfSha256(
    concatBytes(ss1, ss2),
    concatBytes(nonce, nonceResp),
    utf8ToBytes(connectInfo(0)),
    32,
  );
}

/**
 * Ratchet the connection secret to the next epoch with fresh KEM entropy ss3,
 * salted by the rotation ciphertext ct3.
 */
export function rotatedSecret(
  prev: Uint8Array,
  ss3: Uint8Array,
  ct3: Uint8Array,
  epoch: number,
): Uint8Array {
  return hkdfSha256(concatBytes(prev, ss3), ct3, utf8ToBytes(connectInfo(epoch)), 32);
}

/** The private audience secret for an epoch (fed to the AEAD layer). */
export function audienceSecret(connSecret: Uint8Array, epoch: number): Uint8Array {
  return hkdfSha256(connSecret, EMPTY_SALT, utf8ToBytes(audienceInfo(epoch)), 32);
}

/**
 * The private audience code for an epoch. Derived from the secret, so it is
 * unlinkable to either participant's address.
 */
export function audienceCode(connSecret: Uint8Array, epoch: number): Uint8Array {
  return keccak256(audienceSecret(connSecret, epoch)).slice(0, CODE_BYTES);
}
